typing/typemod.ml

```diff
@@ -1786,13 +1774,17 @@ and transl_recmodule_modtypes env sdecls =
   let init =
     List.map2
       (fun id pmd ->
+         let md_uid = Uid.mk ~current_unit:(Env.get_unit_name ()) in
          let md =
            { md_type = approx_modtype approx_env pmd.pmd_type;
              md_loc = pmd.pmd_loc;
              md_attributes = pmd.pmd_attributes;
-             md_uid = Uid.mk ~current_unit:(Env.get_unit_name ()) }
+             md_uid }
+         in
+         let id_shape =
+           Option.map (fun id -> id, Shape.var md_uid id) id
```

Comment:
One issue is that, currently, the approximated signature doesn't contain meaningful uids. So any computation you're going to make on a shape constructed from it is going to be incorrect / useless.

You could of course preallocate uids, and then reuse them for the various iterations of the signature translation.

But I guess the important question is: what should happen of your second example? When asking for the definition of `A.t` should we be brought to line 2 or to line 4?
Our current stance is that it should bring you to line 4, because that's where the definition is. Using a shape constructed from the signature is just not going to get you there. Therefore our placeholder variable shape, that is then looked up during reduction (which happens after the whole recursive block has been typed, and proper shapes computed).

Reply:
Btw, following your comment I looked at this part of the code again, and it looks like it was slightly incorrect: the environment was never updated to contain the proper shapes after we compute them.
@voodoos is going to send a patch fixing that soon (i.e. in the next hour or so).

Reply:
Patch is 80d4c6d

Reply:
Btw: ping @gasche .
tl;dr: the answer to [your question](https://github.com/ocaml/ocaml/pull/10718#issuecomment-954792476) is: no, we use fuel to stop looping forever on your example.

Reply:
To summarize my discussion with @trefis , currently shapes do not support "definition out-of-declaration" where the implementation definition is imported from the signature. This seems like an acceptable compromise since there is no concrete definitions in those instances.